controllers/application/controller.go

```diff
@@ -207,3 +220,32 @@ func (r *ApplicationReconciler) finalizeApplication(ctx context.Context, applica
 	}
 	return err
 }
+
+func (r *ApplicationReconciler) validateApplicationSpec(application *skiperatorv1alpha1.Application) error {
+	validationFunctions := []func(application *skiperatorv1alpha1.Application) error{
+		ValidateIngresses,
+	}
+
+	for _, validationFunction := range validationFunctions {
+		err := validationFunction(application)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func ValidateIngresses(application *skiperatorv1alpha1.Application) error {
+	for _, ingress := range application.Spec.Ingresses {
+		if ingress == "" || util.HasUpperCaseLetter(ingress) || strings.Contains(ingress, " ") {
+			errMessage := fmt.Sprintf("ingress with value '%s' was not valid. ingress must be lower case, contain no spaces, and be a non-empty string", ingress)
+			return errors.NewInvalid(application.GroupVersionKind().GroupKind(), application.Name, field.ErrorList{
+				field.Invalid(field.NewPath("application").Child("spec").Child("ingresses"), application.Spec.Ingresses, errMessage),
```

Comment:
this section can probably be generalised somewhat

Reply:
definitely possible. i'm down to change it to a regex instead of multiple if conditions if that's preferable.

it *is* more readable this way, though.